llvm/lib/Target/EVM/EVMStackLayoutGenerator.cpp

```diff
@@ -398,183 +390,163 @@ void StackLayoutGenerator::processEntryPoint(CFG::BasicBlock const &Entry) {
-              if (It == Layout.blockInfos.end())
-                return Stack{};
-
-              return It->second.entryLayout;
-            }
-            // If the current iteration has already visited the jump target,
-            // start from its entry layout.
-            if (Visited.count(Jump.Target))
-              return Layout.blockInfos.at(Jump.Target).entryLayout;
-            // Otherwise stage the jump target for visit and defer the current
-            // block.
-            ToVisit.emplace_front(Jump.Target);
-            return std::nullopt;
-          },
-          [&](CFG::BasicBlock::ConditionalJump const &ConditionalJump)
-              -> std::optional<Stack> {
-            bool ZeroVisited = Visited.count(ConditionalJump.Zero);
-            bool NonZeroVisited = Visited.count(ConditionalJump.NonZero);
-
-            if (ZeroVisited && NonZeroVisited) {
-              // If the current iteration has already Visited both jump targets,
-              // start from its entry layout.
-              Stack CombonedStack = combineStack(
-                  Layout.blockInfos.at(ConditionalJump.Zero).entryLayout,
-                  Layout.blockInfos.at(ConditionalJump.NonZero).entryLayout);
-              // Additionally, the jump condition has to be at the stack top at
-              // exit.
-              CombonedStack.emplace_back(ConditionalJump.Condition);
-              return CombonedStack;
-            }
+std::optional<Stack> EVMStackLayoutGenerator::getExitLayoutOrStageDependencies(
+    const MachineBasicBlock *Block,
+    const DenseSet<const MachineBasicBlock *> &Visited,
+    std::list<const MachineBasicBlock *> &ToVisit) const {
+  const EVMMBBTerminatorsInfo *TermInfo = CFGInfo.getTerminatorsInfo(Block);
+  MBBExitType ExitType = TermInfo->getExitType();
+  if (ExitType == MBBExitType::UnconditionalBranch) {
+    const MachineBasicBlock *Target = nullptr;
+    bool IsLatch = false;
+    std::tie(std::ignore, Target, IsLatch) = TermInfo->getUnconditionalBranch();
+    if (IsLatch) {
+      // Choose the best currently known entry layout of the jump target
+      // as initial exit. Note that this may not yet be the final
+      // layout.
+      auto It = MBBEntryLayoutMap.find(Target);
+      return It == MBBEntryLayoutMap.end() ? Stack{} : It->second;
+    }
+    // If the current iteration has already visited the jump target,
+    // start from its entry layout.
+    if (Visited.count(Target))
+      return MBBEntryLayoutMap.at(Target);
+    // Otherwise stage the jump target for visit and defer the current
+    // block.
+    ToVisit.emplace_front(Target);
+    return std::nullopt;
+  }
+  if (ExitType == MBBExitType::ConditionalBranch) {
+    const MachineBasicBlock *TrueBB = nullptr, *FalseBB = nullptr;
+    const MachineOperand *Condition = nullptr;
+    std::tie(std::ignore, std::ignore, TrueBB, FalseBB, Condition) =
+        TermInfo->getConditionalBranch();
```

Comment:
```suggestion
    auto [CondBr, UncondBr, TrueBB, FalseBB, Condition] =
        TermInfo->getConditionalBranch();
```
Not sure this is better, since `std::ignore` can't be used in structured bindings.

Reply:
I like the structural bindings here, as it reduces the code. The main reason I not used is to avoid possible 'unused variable'. But OK, it seems the '_' works as  name-independent in our build environment.